doc/Authoring_Plugins.md

```diff
@@ -0,0 +1,188 @@
+
+Theia defines a runtime API allowing plug-ins to customize the IDE and add their behaviour to various aspects of the application. 
+
+In Theia, a plug-in has access to the API through an object named `theia` which is available in all plug-ins.
+[More details on API](https://github.com/theia-ide/theia/blob/master/packages/plugin/API.md).
+
+There are two natures of plug-ins:
+ - Backend plug-in. If you're familiar with VS Code extensions, it's very close. The plug-in's code is running in its own process on the server side. API is called and it's the API that will send some actions on user's browser/UI to register new commands, etc. All the callbacks are executed on the server side on a dedicated process.
+ - Frontend plug-in. In that case, callbacks are executed in a worker thread on the UI/browser. These plug-ins are only authorized to use "browser compliant" modules. For example opening or writing to a file is impossible as all the code of the plug-in is running on the browser side. But this approach is helpful if you really want to have some stuff on the client side to avoid some network operations.
+
+## Prerequisites
+
+Having a running Theia instance. (v0.3.12+)
+Instructions for getting Theia are available from [Theia repository](https://github.com/theia-ide/theia#getting-started).
+
+## Project Layout
+
+We are going to create a new project and for this we will create a folder named `theia-hello-world-plugin` that will contain the source code of the project.
+
+To ease the setup of such a repository, a [Yeoman code generator](https://www.npmjs.com/package/@theia/generator-plugin) exists and can be installed to scaffold the project.
+
+Generator can be installed and executed with the following commands. Note that these commands can be entered from a new terminal inside your running Theia instance.
+
+```bash
+npm install -g yo @theia/generator-plugin
+mkdir theia-hello-world-plugin
+cd theia-hello-world-plugin
+yo @theia/plugin
+```
+
+In the previous commands:
+- `npm install -g yo @theia/generator-plugin` command install globally the Theia generator.
+- `yo @theia/plugin` is calling the yeoman generator asking to use for templating the Theia's plug-in generator.
+
+
+here is animated screenshot of the generator running.
+![Yeoman plugin output](images/plug-ins/yeoman-plugin.gif "Yeoman plugin output")
+
+Pick up default values for each questions.
+
+At this step, in `theia-hello-world-plugin` folder there is a plug-in that is already built with associated source code.
+
+## Implementing the Plug-in
+
+Let's have look at the generated code now.
+
+```json
+{
+      "name": "theia-hello-world-plugin",
+      "publisher": "theia",
+      "keywords": [
+        "theia-plugin"
+      ],
+      "version": "0.0.1",
+      "files": [
+        "src"
+      ],
+      "devDependencies": {
+        "@theia/plugin": "latest", <-- 1. Theia API dependency
```

Comment:
I notice that on my generated package.json, it's picking up the ```next``` version of ```@theia/plugin```, instead of ```latest```. It's also picking up ```"@theia/plugin-packager": "latest"```. Not sure if it's an issue, but mixing next and latest dependencies can be problematic.

Reply:
I will move to latest for both in generated code